src/NHibernate.Test/SystemTransactions/TransactionFixture.cs

```diff
@@ -1,48 +1,280 @@
+				using (var s = OpenSession())
+				{
+					var person = new Person { CreatedAt = createdAt };
+					savedId = s.Save(person);
+
+					if (explicitFlush)
+						s.Flush();
+				}
+				txscope.Complete();
 			}
 
-			using (var session = OpenSession())
+			using (new TransactionScope())
+			{
+				using (var s = OpenSession())
+				{
+					var person = s.Get<Person>(savedId);
+					person.CreatedAt = createdAt.AddMonths(-1);
+
+					if (explicitFlush)
+						s.Flush();
+				}
+
+				// No Complete call for triggering rollback.
+			}
+
+			using (var s = OpenSession())
+			using (s.BeginTransaction())
+			{
+				Assert.AreEqual(createdAt, s.Get<Person>(savedId).CreatedAt, "Entity update was not rollback-ed.");
+			}
+		}
+
+		[Test]
+		public void CanDeleteItem([Values(false, true)] bool explicitFlush)
+		{
+			object id;
 			using (var tx = new TransactionScope())
 			{
-				var w = session.Get<W>(identifier);
-				Assert.IsNotNull(w);
-				// Without explicit Flush, this delay the delete to prepare phase, and test flushing from there
-				// while having already acquired a connection due to the Get.
-				session.Delete(w);
-				tx.Complete();
+				using (var s = OpenSession())
+				{
+					id = s.Save(new Person { CreatedAt = DateTime.Today });
+
+					if (explicitFlush)
+						s.Flush();
+
+					tx.Complete();
+				}
+			}
+
+			using (var s = OpenSession())
+			using (s.BeginTransaction())
+			{
+				Assert.AreEqual(1, s.Query<Person>().Count(), "Entity not found in database.");
+			}
+
+			using (var tx = new TransactionScope())
+			{
+				using (var s = OpenSession())
+				{
+					s.Delete(s.Get<Person>(id));
+
+					if (explicitFlush)
+						s.Flush();
+
+					tx.Complete();
+				}
+			}
+
+			AssertNoPersons();
+		}
+
+		[Test]
+		public void CanUseSessionWithManyScopes([Values(false, true)] bool explicitFlush)
+		{
+			using (var s = Sfi.WithOptions().ConnectionReleaseMode(ConnectionReleaseMode.OnClose).OpenSession())
+			{
+				using (var tx = new TransactionScope())
+				{
+					// Acquire the connection
+					var count = s.Query<Person>().Count();
+					Assert.That(count, Is.EqualTo(0), "Unexpected initial entity count.");
+					tx.Complete();
+				}
+
+				using (var tx = new TransactionScope())
+				{
+					s.Save(new Person { CreatedAt = DateTime.Today });
+
+					if (explicitFlush)
+						s.Flush();
+
+					tx.Complete();
+				}
+
+				using (var tx = new TransactionScope())
+				{
+					var count = s.Query<Person>().Count();
+					Assert.That(count, Is.EqualTo(1), "Unexpected entity count after committed insert.");
+					tx.Complete();
+				}
+
+				using (new TransactionScope())
+				{
+					s.Save(new Person { CreatedAt = DateTime.Today });
+
+					if (explicitFlush)
+						s.Flush();
+
+					// No complete for rollback-ing.
+				}
+
+				// Do not reuse the session after a rollback, its state does not allow it.
+				// http://nhibernate.info/doc/nhibernate-reference/manipulatingdata.html#manipulatingdata-endingsession-commit
```

Comment:
I had forgotten that myself, causing the check to fail because the count was flushing the "rollback-ed" insert if the flush was not explicitly done.